### PR TITLE
feat: expand color palette flexibility

### LINKED ISSUE #1062: expand `colorPalette` flexibility
### Description

expand the flexibility of the Virtual Color feature by allowing for un-opinionated nested object definitions

### Problem Statement/Justification

Hi! I've been using panda in two large projects. As a former Chakra user and a tailwind tolerator, I was delighted to hop on the panda train as quickly as possible and guide my teams in that direction.

One idea I really like is [Virtual Color](https://panda-css.com/docs/guides/virtual-color).

I like to build component libraries in a very flexible way, and a technique I use is to provide `aesthetic` and `color` props. In panda, I model it like this:

```ts
const config = defineConfig({
	theme: {
		tokens: {
			blue: {
				50: {
					value: 'blue'
				},
				900: {
					value: 'cyan'
				}
			},
			red: {
				50: {
					value: 'red'
				},
				900: {
					value: 'magenta'
				}
			}
		}
	}
});

const buttonRecipe = defineRecipe({
	name: 'button',
	base: {
		…baseProps
	},
	variants: {
		color: {
			blue: {
				colorPalette: 'blue',
			},
			red: {
				colorPalette: 'red',
			}
		},
		aesthetic: {
			solid: {
				backgroundColor: 'colorPalette.50'
				color: 'colorPalette.900'
			}
		}
	}
});
```

This is straightforward enough.

However, in my current project, we're taking an approach where we lean into semantic tokens, with a theme configuration like so:

```ts
const config = defineConfig({
	theme: {
		semanticTokens: {
			core: {
				light: {
					DEFAULT: {
						value: 'blue'
					},
					accent: {
						value: 'cyan'
					}
			},
			brand: {
				light: {
					DEFAULT: {
						value: 'red',
					},
					accent: {
						value: 'magenta'
					}
				}
			}
		}
	}
})
```

which means I want to define my button like this:

```ts
const buttonRecipe = defineRecipe({
	name: 'button',
	base: {
		…baseProps
	},
	variants: {
		color: {
			core: {
				colorPalette: 'core',
			},
			brand: {
				colorPalette: 'brand',
			}
		},
		aesthetic: {
			solid: {
				backgroundColor: 'colorPalette.light'
				color: 'colorPalette.light.accent'
			}
		}
	}
});
```

This doesn't work.

### Proposed Solution or API

I see [in the docs that `colorPalette` requires 50–900 keys](https://panda-css.com/docs/guides/virtual-color#:~:text=%F0%9F%92%A1-,Important%3A%20A%20valid%20colorPalette%20value%20is%20an%20color%20object%20that%20has%2050%20to%20900%20keys.,-const%20className%20%3D).

1. Why does `colorPalette` have an opinion about its shape?
2. Would it be reasonable to expand the flexibility of Virtual Colors to allow for nested tokens?

### Alternatives

- altering my color palette to use `50`-`900`
- not using `colorPalette` or Virtual Color at all

### Additional Information

Here's a failing case: https://play.panda-css.com/XIIQGvJhd9

Thanks for the excellent work on the project – it makes building UIs so fun, and is the most lovely successor to chakra, which I loved for a long time.

### PR DESCRIPTION
**Pull request by isBatak**
Closes #1062 issue
Related to #1103 discussion

## 📝 Description

> Add a brief description

## ⛳️ Current behavior (updates)

> Please describe the current behavior that you are modifying

## 🚀 New behavior

> Please describe the behavior or changes this PR adds

### CODE REVIEW

File: packages/token-dictionary/src/middleware.ts [hunk 1]
```diff
@@ -71,6 +71,15 @@ export const addVirtualPalette: TokenMiddleware = {
...
       const { colorPalette } = token.extensions
       if (!colorPalette) return
+      const colorPaletteRoot = colorPalette.split('.').shift()
+      token.path
+        .slice(token.path.indexOf(colorPaletteRoot) + 1)
+        .reduce((acc, _, i, arr) => {
+          acc.push(arr.slice(0, i + 1).join('.'))
+          return acc
+        }, [] as string[])
+        .forEach((key) => keys.add(key))
```
Comment (astahmer) [inline packages/token-dictionary/src/middleware.ts:81]: could you add a little comment above so we know with a glance what's happening here ?

Comment (isBatak) [reply packages/token-dictionary/src/middleware.ts]: Ofc I'll do an refactoring and commenting phase, this is just a PoC for now.
I'm still worried about the performance and some memory leaks.

Comment (isBatak) [reply packages/token-dictionary/src/middleware.ts]: For example, now `colorPalettes` have a lot of duplicate tokens
image

**Comment by segunadebayo** [comment]
What’s the status of this @isBatak?

**Comment by segunadebayo** [comment]
I'm closing this for now. Kindly re-open when you have a working POC or ready for review.

> On a quick glance, it feels a bit too complex for the benefit it brings, but I could be wrong.

File: packages/token-dictionary/src/middleware.ts (hunk 1 above)
Comment (isBatak) [reply packages/token-dictionary/src/middleware.ts]: Duplicates are now fixed 
image

**Comment by isBatak** [comment]
@segunadebayo it's ready for review

**Review by astahmer — APPROVED** [review]

File: .changeset/late-carrots-grab.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@pandacss/token-dictionary': minor
+---
+
+Fix for issue 1062: expand colorPalette flexibility
```
[the remark this answers is not in the document]
Comment (isBatak) [reply .changeset/late-carrots-grab.md]: @anubra266 done!

File: .changeset/late-carrots-grab.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
+---
+'@pandacss/token-dictionary': minor
+---
+
+- Fix for issue 1062: expand `colorPalette` flexibility
```
Comment (anubra266) [inline .changeset/late-carrots-grab.md:5]:
```suggestion
```

**Review by anubra266 — APPROVED** [review]

**Review by anubra266 — APPROVED** [review]

**Pull request merged**